### ISSUE TITLE
Probably a mistake in BatUref.unite, performance impact

### ISSUE THREAD

**Issue by bjeannet** [ocaml-batteries-team/batteries-included#771]
batUref.ml is implementing Union-find algorithm with ranks, the rank of a root being the maximum possible height of the tree attached to it.
Looking at the code of the unite function, it seems to me that the policy related to the ranks is wrong and should be inverted. 
- According to wikipedia (and confirmed by my intuition), the root "r1" of the shortest tree (smallest rank) should point to the root "r2" of the tallest tree (greatest rank);  as a result the rank of the new root "r2" does not change (unless the ranks are equal, in which case it is incremented by one).
- What I understand is that the code performs the reverse, that is, the root of the tallest tree points to the root of the shortest tree (which should imply that the rank of the new root should always be the rank of the tallest tree plus one, which is not the case in the code)

**Comment by gasche** [comment]
I agree with your reading of the code, but would be reassured by a test. Could you take out the implementation of the batUref as an independent file, and write a `rank` function, and create a sequence of unification calls that reveals the issue and confirms that the fixes behave as expected?

**Comment by bjeannet** [comment]
I'll do it, but not before monday (one test showing that currently computed rank is wrong, and one test on the fixed version showing that (1) rank is OK (2) stays 1 or 2 on a common sequence of calls  to unite.
 
I point out that functionally the current code is OK, only performances are impacted: cost of N calls to unite could be in O(N*N) instead of in O(N\alpha(N)).
